procfs_linux.go

```diff
@@ -0,0 +1,105 @@
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"sync"
+
+	"golang.org/x/sys/unix"
+)
+
+var (
+	procSelfFdHandle *os.File
+	procSelfFdError  error
+	procSelfFdOnce   sync.Once
+
+	errUnsafeProcfs = errors.New("unsafe procfs detected")
+)
+
+func doGetProcSelfFd() (*os.File, error) {
+	procRoot, err := os.OpenFile("/proc", unix.O_NOFOLLOW|unix.O_PATH|unix.O_CLOEXEC, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Use fsopen or open_tree to get a safe handle that cannot be
+	// over-mounted and we can absolutely verify.
+
+	// Make sure that the /proc handle is actually the root (f_type is
+	// PROC_SUPER_MAGIC, and st_ino is PROC_ROOT_INO).
+	const (
+		PROC_SUPER_MAGIC = 0x9fa0
+		PROC_ROOT_INO    = 1
+	)
+
+	var stat unix.Stat_t
+	if err := unix.Fstat(int(procRoot.Fd()), &stat); err != nil {
+		return nil, &os.PathError{Op: "fstat", Path: procRoot.Name(), Err: err}
+	}
+	if stat.Ino != PROC_ROOT_INO {
+		return nil, fmt.Errorf("%w: incorrect procfs root inode number %d", errUnsafeProcfs, stat.Ino)
+	}
+
+	var statfs unix.Statfs_t
+	if err := unix.Fstatfs(int(procRoot.Fd()), &statfs); err != nil {
+		return nil, &os.PathError{Op: "fstatfs", Path: procRoot.Name(), Err: err}
+	}
+	if statfs.Type != PROC_SUPER_MAGIC {
+		return nil, fmt.Errorf("%w: incorrect procfs root filesystem type 0x%x", errUnsafeProcfs, statfs.Type)
+	}
+
+	handle, err := openatFile(procRoot, "self/fd", unix.O_PATH|unix.O_NOFOLLOW|unix.O_DIRECTORY|unix.O_CLOEXEC, 0)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", errUnsafeProcfs, err)
+	}
+	// We can't detect bind-mounts of different parts of procfs on top of
+	// /proc (a-la RESOLVE_NO_XDEV), but we can at least be sure that we
+	// aren't on the wrong filesystem here.
+	if err := unix.Fstatfs(int(handle.Fd()), &statfs); err != nil {
+		return nil, &os.PathError{Op: "fstatfs", Path: handle.Name(), Err: err}
+	}
+	if statfs.Type != PROC_SUPER_MAGIC {
+		return nil, fmt.Errorf("%w: incorrect /proc/self/fd filesystem type 0x%x", errUnsafeProcfs, statfs.Type)
+	}
+	return handle, nil
+}
+
+func getProcSelfFd() (*os.File, error) {
+	procSelfFdOnce.Do(func() {
+		procSelfFdHandle, procSelfFdError = doGetProcSelfFd()
+	})
+	return procSelfFdHandle, procSelfFdError
+}
+
+func procSelfFdReadlink(f *os.File) (string, error) {
+	// NOTE: It is possible for an attacker to bind-mount on top of the
+	// /proc/self/fd/... symlink, and there is currently no way for us to
+	// detect this. So we just have to assume that hasn't happened...
+	procSelfFd, err := getProcSelfFd()
+	if err != nil {
+		return "", fmt.Errorf("get safe procfs handle: %w", err)
+	}
+	// readlinkat(</proc/self/fd>, "42")
+	return readlinkatFile(procSelfFd, strconv.Itoa(int(f.Fd())))
+}
+
+var errPossibleBreakout = errors.New("possible breakout detected")
+
+func checkProcSelfFdPath(path string, file *os.File) error {
+	actualPath, err := procSelfFdReadlink(file)
+	if err != nil {
+		return fmt.Errorf("get path of handle: %w", err)
+	}
+	if actualPath != path {
```

Comment:
Checking for `actualPath != path` does not work when `root` refers to a foreign mount such as:

```go
securejoin.MkdirAll("/proc/1234/root", "/home/alice", 0o755)
```

Example how to test this:
* Start a container in terminal 1
```
$ docker run -ti --rm --name mycontainer1 ubuntu
# mkfifo /home/fifo
# sleep 1000 > /home/fifo
```

* Start a privileged container in terminal 2
```
$ MYCONTAINER1_PID=$(docker inspect mycontainer1 -f json|jq '.[0].State.Pid')
$ docker run --privileged -ti --rm --name myprivilegedagent -v /proc:/host/proc -e MYCONTAINER1_PID=$MYCONTAINER1_PID ubuntu

# cat  /host/proc/$MYCONTAINER1_PID/root/home/fifo &
# readlink /proc/$(pidof cat)/fd/3
/home/fifo
```

Here the `readlink()` returns the `actualPath` string "/home/fifo" that cannot be meaningfully compared to `/host/proc/81550/root/home/fifo`.

Additionally, the kernel might append the string `" (deleted)"` in an ambiguous way:
https://github.com/torvalds/linux/blob/v6.9/fs/d_path.c#L256

Reply:
We could fetch the "real" path of the root through `/proc/self/fd` and compare against that. This is what libpathrs does, I wasn't sure if it was necessary (in libpathrs, we use a custom structure instead of `*os.File` so caching it is nicer).

That being said, I'm not sure how I feel about doing "safe" path operations through magic-links. With `openat2` it can be done safely, but `/proc/self/fd` can give really weird results for paths that are not in your tree and IMHO it's better to error out in some of these cases...

> Additionally, the kernel might append the string " (deleted)" in an ambiguous way:

AFAIK it should not be possible to get ` (deleted)` for a directory, because `rmdir` fails if there is any open handle to the directory.

Reply:
I've pushed a version which saves the readlink result, but I'll leave this open for discussion about ` (deleted)`.

Reply:
Okay, I've now pushed a patch to check for dead directories when we do `checkProcSelfFdPath` as well as a sanity-check in `MkdirAll`.